### DISCUSSION TITLE
difference between ormdantic and sqlmodel?

### DISCUSSION THREAD

**Discussion by 0xsirsaif** [yezz123/ormdantic#25] [Q&A] [answered]
Great! but could you show the key difference between this library and `sqlmodel`?

**Comment by iron3oxide** [comment, marked as answer]
This should probably be a discussion instead, but I'll give it a try (disclaimer: I have not spent too much time reading through both code bases, so this is not very detailed and I might be wrong in some aspects). These are the main differences I can see after a brief study:

1. `ormdantic` uses a wrapper/decorator to make pydantic BaseModels behave like database tables as well, whereas `sqlmodel` inherits from pydantic BaseModel and uses a modified SQLAlchemy base model as its metaclass to achieve the same thing.
2. In `sqlmodel`, the database connection is not built into the main class (SQLModel), but handled by separate equivalents of SQLAlchemy classes and functions (Session and create_engine).
Since the async equivalents are not available in `sqlmodel` yet, one has to use the original SQLAlchemy ext versions, which is certainly possible, but imho a bit cumbersome until SQLAlchemy 2.0 is released (type checking doesn't really work for me atm). 
`ormdantic` takes the engine as a constructor parameter instead and creates sessions internally where needed. It is "async by default" one might say.

3. The aforementioned difference leads to another difference: in sqlmodel, you typically interact with both a Session object and a SQLModel object, e.g.:
    ```python
    ...
    hero = Hero()   # Hero inherits from SQLModel
    session.add(hero)
    await session.commit()
    ...
    ```
    In `ormdantic`, you just interact with the database abstraction:
    ```python
    ...
    hero = Hero()   # Hero inherits from pydantic BaseModel and is decorated as database table
    await database["Hero"].insert(hero)
    ...
    ```
    I find the latter to be a bit more elegant/intuitive, but that's of course subjective and probably because I am used to writing SQL.

4. `ormdantic` adds pypika to the mix, `sqlmodel` does not seem to include it or anything similar, but that's probably an internal difference. It is an added dependency though for those who care about it.
5. Another difference is of course that `sqlmodel` is a bit more established and has the benefit of being authored by the creator of FastAPI, so on paper it should be the ORM to use with FastAPI. Unfortunately said creator has a lot on his plate and `sqlmodel` hasn't seen a new release for some time now. `ormdantic` on the other hand is of course brand new and thus unproven, but could be an exciting alternative imho.

**Comment by yezz123** [comment]
Well, Thank you @iron3oxide for the Amazing explanation for both `SQLModel` and `Ormdantic` 🙏🏻 

I will add a few things for you @0xsirsaif , is ormdantic similar to PonyORM a (Pure python queries) and that's mean the ability to write queries to the database using Python generator expressions and lambdas, Its make the close structure to writing SQL Queries but making it more easy and easy for new people to integrate it!.

**Comment by yezz123** [comment]
Hey @iron3oxide I add your Comment to the docs, cause I find it more clear and more good comparison between the two packages!

https://ormdantic.yezz.me/faq/#difference-between-ormdantic-and-sqlmodel

Thank you so much 🙇🏻‍♂️

**Reply by iron3oxide** [reply]
No problem! One small caveat though: I just noticed that my comment about ormdantic taking an engine as an argument is out of date since #18 changed it to use a connection string instead. Should I open a PR for that?

**Reply by yezz123** [reply]
> No problem! One small caveat though: I just noticed that my comment about ormdantic taking an engine as an argument is out of date since #18 changed it to use a connection string instead. Should I open a PR for that?

Yeah sure, thank you so much 🙏🏻
